### PR TITLE
Extensive models & Wrapup

### PR DESCRIPTION
**Pull request by DaluS**
# The big PR 

This Pull Request will contains most of last month advances, since I was alone on this segment, and since it is composed of increments here and there, they are all together. Little has been done on the core itself (which I have difficulties understanding and modifying), So it's a sum of smaller contributions. 

This PR is based on #112 

# ISSUES

A lot of small (or less small) issues still exist in the core as it is :

Here is a list of issues that has been observed : 

## Python routing

Some elements are generated the first time a terminal is 

## Model files

1. We cannot define parameters inside the model file itself 
2. `time` cannot be called using scipy solvers #115
3. plot in preset is still not implemented
4. There is no reason for `_FUNC_ORDER` to be in the file as it is overriden anyway 

## Hub load and presets

1. When creating a hub with a model and a preset, the model is kinda loaded twice with something unclean on undeclared parameters _However the preset is working as intended_
image
2. `set_dparam` works well for one value but erase everything in the case of a dictionnary
image
image
3. Iterating over a dictionary works almost but for grid which is written with another formalism in both cases ( if in a dictionary `("key" : {'value':vector,'grid':True})`, or in a unique parameter `(key="key" ,value=vector,grid=True)`,
4. dpreset works well ! Day-to-day use is now creating a dictionary, putting it inside dpreset with a name, and creating the hub with dpreset and the name of the preset.  

## Solvers 

5. I have found no way to use scipy solvers on constant time-step, making it difficult to verify that the way we write our system is indeed working...
6. The Rk4 does not behave as a Rk4 ? It is unclear
7. Scipy solvers are not functioning in parallel

## Plots 

This is clearly minor ! 

8. Linestyle nor linewidth taken into account in hub.plot
9. `hub.plot` just to plot is not 

## Misc

* The documentation is out of date (doc folder, readme)
* Requirements should be updated too

# Modifications 

## Small modifications 

* Now the structure has a folder "example" that contains cockpit-style files. It accelerate the use of the library in general. `AllFormulations.py` contains all the commands with typical use 
* Lot of files have been displaced to the NotIncluded Folder, and should not be reviewed
* Def fields have now carbon concentration as a unit `C`, a `temporary` group in the beginning of `_LIBRARY` because of specific parameters to one model
*  `` means dimensionless, and `None` is the default (easier to code that way given the existing structure)
* `n` has the role `beta` had before in goodwin-keen type of models (population growth), to allow `beta` being something else
* Reorganisation of the file and addition of climate-related parameters and next-gen models
* Back-up moved to `NotIncluded`
* An rk1 solver (Euler scheme) has been added for test purpose
* All models name having `_` has been change into `-`
* Models not fully functional are in `_models\NotWorkingYet`

# Models added

## Non-economic model

We added for control purpose and for future coupling some simple ODE models : 
* The 3-layer climate model from GEMMES model
* A double pendulum for chaotic motions under constraints
* Lorenz Chaotic system/butterfly
* A second order dampened oscillator for numerical/theoretical comparison

## Economic model

`GK` Has been widely scrutinized and corrected ! Now gives the same results as GK-Reduced, so we can safely construct on them.

A few others has been added and compared to each others

# Added tools 

These has been put in the `__init__` so that they can be called directly through pygemmes 

## Model preset copy 

One might want to compare a Full model to its reduced version, they have the same logic behind but the ode are not on the same terms. This module take a preset from an extensive model and creates its equivalent (mostly by the calculation of the statevar that correspond) for the reduced model. 

Typically using 
`pgm.create_preset_from_model_preset('GK','GK-Reduced',                                    targetpreset=False,targetdpreset=False,returnas='hub')`

## Solver comparison 

`plot_one_run_all_solvers(_MODEL, preset=False, _DPRESET=False)`

Allow to see, given the conditions of the preset, how important the solver is on the characteristics of the run

## Solver comparison on Lorenz

Same as above but giving a nice 3D visualisation !

## testConvergence_DampOsc 

Gives the difference for the solver you want to test between numerical and theoretical solution 

`pgm.testConvergence_DampOsc(np.logspace(-1, -4, 7), 'eRK4-homemade')`

## in models/__init__.py 

I commented line 96 to 100 as it was not working (preset print in get available models)

## Solvers scipy sometimes diverge with no reasons... 

Testing the double pendulum on all solvers gives : 
image

So not all of them are giving a solution, we can understand that the rk1 is not working however it is more surprising for scipy-rk4. It also happens on normal run with phase transitions/bifurcation. I do not master enough these solvers to avoid that problem...

### CODE REVIEW

**Comment by DaluS** [comment]
Hi @Didou09 , I think this should give a list of explicit improvement, modifications we can bring ! I'm curious if you have time to discuss about the project, your help would be very valuable !

**Comment by Didou09** [comment]
From what I understand, the 9 commits from PR #112 are included as the first 9 commits of this PR.
So it appears that the previous PR is not necessary as it is included in this one.

I will thus resume the review process here starting from the commit 10.

Here is what I concluded at the end of the review process of PR #112: 

> 
> I am ok with this commit, but there are still 2 remaining caveats: test_02_Hub_Multiple.py and non-regression tests.
> 
>     Unit tests are not passing, so with your approval @DaluS I will fix unit tests and solve the 2 remaining caveats

**Review by Didou09 — COMMENTED** [review]
Commit 10:    #595c8943560332f98acb37361f7cb480bcec2db0
----------------

* Overlooking `Examples/BASIN-ATTRACTION-GK.py`, `Examples/Cockpit.py`, `Examples/Cockpit_COMPARING-REDUCED-NOT-REDUCED.py`, `Examples/Draw_NetworkX_Graph.py`, `Examples/TEST-CONVERGENCE.py`, `NotIncluded/_def_fields_backup.py`, 

* `_def_fields.py`: ok, (added 2 symbols and removed deprecated dict)

* `_model_GK-Reduced.py`: ok, minor changes to equations + renamed `i` to `inflation`

* `_model_GK.py`: ok, this model was necessary for unit tests, this commit fixes one of the 3 caveat from previous commits

**Review by Didou09 — COMMENTED** [review]
Commit 11:  #97d72c918c359a0f63e6712d8b0b8eb4eecec782
----------------

* Overlooking `Examples/Cockpit-MonoGEM.py`, `Examples/Cockpit.py`, `Examples/Draw_NetworkX_Graph.py`

* `_def_fields.py`: ok (minor renaming of 2 parameters: `fC` and `Omega0`)

* `_model_MonoGEM.py`: ok, model back (was deleted some commits ago)

* `_model_MonoGEM_Reduced.py`: ok, brand new model

I am ok with this commit, the previous caveats are remaining

**Review by Didou09 — COMMENTED** [review]
Commit 12: #8bb08dca499bfe2029e79e629c1bec9e0d727e6d
----------------

* Overlooking `Examples/Cockpit_COMPARING-REDUCED-NOT-REDUCED.py`, `Examples/Draw_NetworkX_Graph.py`

* `_def_fields.py`: ok, renamed `i` as `inflation`

* `_model_GK-Reduced.py`: ok, minor changes to equations (`lambda` and `omega`)

**Review by Didou09 — COMMENTED** [review]
Commit 13:    #437c38a1f
---------------

* Overlooking `Examples/Cockpit_COMPARING-REDUCED-NOT-REDUCED.py`

I am ok with this commit, the 2 previous caveats are remaining

**Review by Didou09 — COMMENTED** [review]
Commit 14:    #70822f882279
---------------

* `_model_DoublePendulum.py`:  ok, brand new model (good idea)

I am ok with this commit, the previous 2 caveats are remaining

**Review by Didou09 — COMMENTED** [review]
Commit 15:      #0ffc7b989378de
----------------

* Overlooking `Examples/AllFormulations.py`, `Examples/BASIN-ATTRACTION-GK.py`, `Examples/Cockpit.py`

* Moving of several models to `NotIncluded/`: ok (`DampOscillator`, `DoublePendulum`, `LorenzSystem`, `MonoGEM`, `MonoGEM_Reduced`)

* `_model_3Layerclimate.py`: ok, brand new model

* `_core.py`: ok, minor formatting

* `_solvers.py`: ok, minor formatting

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 16:     #39c28c80bc71535
----------------

* `_model_3Layerclimate.py`: ok (minor parameter value change)

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 17:       #48b471c4c5ffab
----------------

* Overlooking all files in `Examples/`, and `mygif.gif`

* `_def_fields.py`: ok (added parameters for double-pendulum and for 3 layers climate models)

* `_model_3Layerclimate.py`: ok, minor changes to equations and presets

* `_model_DampOscillator.py`: ok, model back in use (was removed by previous commits)

* `_model_DoublePendulum.py`: ok, model back in use

* `_model_GK-Reduced.py`: ok, mionor changes to equations and presets

* `_model_GK.py`: ok, minor reformatting

* `_model_LorenzSystem.py`: ok, model back in use

* `_model_MonoGEM.py`: ok, model back in use

* `_plots/matplotlibstyle.mplstyle`: ok, plot styling parameters

* `_plots/matplotlibstyle.mplstyle.py:` ok, harmless

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 18:         #3a154c3aef018b
----------------

* Overlooking file sin `Examples/`

* `_model_GK.py`: ok, minor reformatting and updated parameter names

* `_model_Goodwin-Reduced.py`: ok, model back in use

* `_model_Goodwin.py`: ok, model back in use

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 19:       #aaf1a945f7222249
----------------

* Overlooking  files in `Examples/`

* `_model_GK.py`: ok, changes to presets

* `_model_Goodwin.py`: ok, minor parameter value update

I am ok with this commit the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 20:     #3c7874f57dc1127a
---------------

* `__init__.py`: ok but bad practice, the function is probably functional, but:
    * this file is typically used for package importing or for things that cannot be done elsewhere
    * this should be coded in a dedicated file 
    * actually this particular functionality could be even implemented in the Hub class itself (later low-priority task)

I am ok with this commit as it does not seem to break anything, but it should probably be corrected in a later dedicated issue / PR for better practice (to be discussed later)

The previous 2 caveats remain

File: pygemmes/_utilities/_solvers.py [hunk 1]
```diff
@@ -192,6 +192,28 @@ def solve(
...
+    if solver == 'eRK1-homemade':
+        _eRK1_homemade(
+            y0=y0,
+            dydt_func=dydt_func,
+            dparam=dparam,
+            lode=lode_solve,
+            lstate=lstate,
+            nt=nt,
+            dverb=dverb,
+        )
```
Comment (Didou09) [inline pygemmes/_utilities/_solvers.py:210]: Redundant code (repeated if statement?)

**Review by Didou09 — COMMENTED** [review]
Commit 21:    #5eda3a2851ac
----------------

* Overlooking all files in `Examples/`, and in `Depreciated/`

* `__init__.py`: ok, same comment as before a new routine is added that is probably useful and operational but bad practice to put it in here

* _model_GK-Reduced.py: ok, minor reformatting

* `_solvers.py`: ok, rk1 added, but there seems to be a repetition of the if statement (see below)

I am ok with this commit, there are now 3 caveats (2 previous + 1 new):
* `test_02_Hub_Multiple.py` needed
* data for non-regression tests needed
* `_solvers.py`: Repeated if statement for rk1

**Review by Didou09 — COMMENTED** [review]
Commit 22:   #907309c83a35
---------------

* Overlooking files in `Examples/`

* `__init__.py`: ok, minor changes

* `_solvers.py`: ok, fixed caveat from previous commit (redundant `if statement`)

I am ok with this commit, the 2 previous caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 23:      #8805acb567bed
----------------

* Overlooking files in `Examples/`

* `__init__.py`: ok, same as for previous commits, it works but it is not good practice in the long term

* `_def_fields.py`: ok, added parameter `beta`

I am ok with this commit, 2 previous caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 24:       #0989452313
---------------

* Overlooking files in `Examples/`

* `__init__.py`: ok, same as before, bad practice

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 25:        #b79cf4d1c2d0ebbc1
----------------

* Overlooking files in `Examples/`

I am ok with this commit, previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 26:      #a87c9090676
----------------

* Overlooking files in `Examples/`

* `__init__.py`: ok, minor changes

I am ok with this commit, previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 27:    #aa203c5aecbf52
---------------

* Overlooking files in `Examples/` and `subgraph.png`

* `__init__.py`: ok (showVariableGraph() added, ok but bad practice)

* `_core.py`: ok, added prints to `Hub.FillCyclesForAll()` and `Hub.FillCycles()`, presumably for debugging

* `_plots/_plots.py`: ok, added prints for debugging

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 28:       #a8fd20e8df163
----------------

* Overlooking files in `Examples/`, `mygif.gif`, `subgraph.png`

* `_core.py`: not ok, this commit pastes a large part of `Hub` methods in itself, duplicating several methods (including `run()`)

* `_def_fields.py`: ok, added group `GK-Improvements`

* `_model_GK-Buffers.py`: ok, brand new model

* `_plots/_plot_timetraces2.py`: ok, but looks like a copy-paste of the original except for 1 more routine at the end, why not just import the original and put only the additional routine in this file?

I am not ok with this commit, there are now 3 caveats (2 previous +1 new):
* `test_02_Hub_Multiple.py` needed
* data for non-regression tests needed
* `_core.py`: a part of `Hub` methods have been copy-pasted in itself (lines 448 - 1251)

**Review by Didou09 — COMMENTED** [review]
Commit 29:      #6bde471103ffa510
----------------

* Overlooking files in `Examples/`

* `_def_fields.py`: ok, minor parameter renaming

* `_model_GK-Buffers.py`: ok, removed `lambdaperciev` buffer + minor change of equations

* `_model_Goodwin.py`: ok, minor change to equations / parameters

* `tests/test_02_Hub_Multiple.py`: ok, back in use, removes a caveat

I am ok with this commit, there are now only 2 caveats left:
* data for non-regression tests needed
* `_core.py`: a part of `Hub` methods have been copy-pasted in itself (lines 448 - 1251)

**Review by Didou09 — COMMENTED** [review]
Commit 30:    #555f9b65dbf
----------------

* Overlooked files in `NotIncluded/`

I am ok with this commit, previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 31:    #bf67a48d5432
---------------

* Overlooked `.pep8speaks.yml`

I am ok with this commit, previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 32:     #3c9f5d34f1215
----------------

* Overlooked files in `Examples/`

* `__init__.py`: ok, but removes presets from `get_available_models()`

* Removed several models: `DoublePendulum`, `GK-Buffer`, `Goodwin`, `MonoGEM`

* `test_02_Hub_Multiple.py`: ok, corrected name of model `G-Reduced`

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 33:    #b92e84601f1dcc
----------------

* Moved several models in `_models/NotWorkingYet/ `  => overlooking

I am ok with this commit the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 34:   merging commit
---------------

**Review by Didou09 — COMMENTED** [review]
Commit 35: merging commit
---------------

**Review by Didou09 — COMMENTED** [review]
Commit 36:      #25769e8c4f4cd
---------------

* `test_02_Hub_Multiple.py`: ok, minor typo (model name)

I am ok with this commit, the previous 2 caveats remain

**Review by Didou09 — COMMENTED** [review]
Commit 37:      merge commit
--------------

**Comment by Didou09** [comment]
On the whole this PR leaves 2 caveats to be soled before merging:
* data for non-regression tests needed
* `_core.py`: a part of `Hub` methods have been copy-pasted in itself (lines 448 - 1251)

I suspect some of them mght be sovled by the next PR, so I will go on t review the next PR before merging anything

**Comment by Didou09** [comment]
All susequent commits were validated in PR #124

**Comment by DaluS** [comment]
> All susequent commits were validated in PR #124

Awesome, I close this PR in consequence

**Comment by Didou09** [comment]
> > All susequent commits were validated in PR #124
> 
> Awesome, I close this PR in consequence

No no no @DaluS , this PR must not be **_closed_**, it must be _**merged**_, but before merging it I need to remove the last caveat, which I will do tomrrow morning  :-)

**Comment by Didou09** [comment]
So, it looks like we still have a regression issue:

image

It concerns the following models_presets_solvers:
* LorenzSystem_Canonical_eRK8-scipy
* DampOscillator_Perfect_eRK2-scipy
* LorenzSystem_Canonical_eRK4-scipy


My comments:
--------------------

* We notice the non-regression appears on Ubuntu, and the original data was produced on my pc which is also ubuntu, so it does not seem to be a platform-issue. It might be a scipy-compilation issue, or something else.
* The good thing is:
     * the regression was detected, so we can investigate it
     * the regression does not seem to affect other models_presets_solvers

Current investigations:
------------------------------

* My pc uses Python 3.7.6, with scipy 1.6.2, while the GA unit tests robot uses Python 3.7.12, with scipy 1.7.3, this might be an explanation.

**Comment by Didou09** [comment]
Update:
-----------

I've tried producing the reference data for non-regression the same environment as GA to rule out version issues, the problem persists.

image


It means it is due to something else, compilation ? 

Suggested workaround:
--------------------------------

I suggest we relax the constraint in these 3 particular cases, this way we keep it in mind (I've opened issue #128 to keep track of this) but allow the merging process to proceed since none of the key models for economy are affected.

**Comment by Didou09** [comment]
The workaround works, it confirms that for some reason only the 3 identified model_preset_solver fail, for all platforms
Further investigations shall be conducted in the framework of issue #128

**Comment by Didou09** [comment]
> It's not surprising that Lorenz is giving non-regression : it's the epitome of chaotic systems and thus individual trajectories are profoundly unstable (every noise is exponentially growing).
> The correct way to test it would be to look how often one point is in the basin of one or the other attractor.
> The same thing will happens for the double pendulum !

Yes, but what is surprising is that is happens only for specific solvers... and that the discrepancies with the reference are always the same...
This suggests that there might something else that we are not seeing

**Pull request merged**